mcs/class/System/Mono.AppleTls/AppleTlsContext.cs

```diff
@@ -684,18 +685,23 @@ public SecTrust GetPeerTrust (bool requireTrust)
 		[Mono.Util.MonoPInvokeCallback (typeof (SslReadFunc))]
 		static SslStatus NativeReadCallback (IntPtr ptr, IntPtr data, ref IntPtr dataLength)
 		{
-			var handle = GCHandle.FromIntPtr (ptr);
-			if (!handle.IsAllocated)
-				return SslStatus.Internal;
+			AppleTlsContext context = null;
+			try {
+				var weakHandle = GCHandle.FromIntPtr (ptr);
+				if (!weakHandle.IsAllocated)
+					return SslStatus.Internal;
 
-			var context = (AppleTlsContext) handle.Target;
-			if (context.disposed)
-				return SslStatus.ClosedAbort;
+				var weakReference = (WeakReference) weakHandle.Target;
+				if (!weakReference.IsAlive)
+					return SslStatus.ClosedAbort;
+
+				context = (AppleTlsContext) weakReference.Target;
+				if (context == null || context.disposed)
+					return SslStatus.ClosedAbort;
 
-			try {
 				return context.NativeReadCallback (data, ref dataLength);
 			} catch (Exception ex) {
-				if (context.lastException == null)
+				if (context != null && context.lastException == null)
```

Comment:
Not if `GCHandle.FromIntPtr` raises an exception.   My understanding is that Martin added that after someone from the runtime team asked for that, the original patch did not include it:

https://gist.github.com/migueldeicaza/83edb2f9e80bdc9fb34d91a9068dc80d